plugin/trino-iceberg/src/main/java/org/apache/iceberg/snowflake/TrinoIcebergSnowflakeCatalogFactory.java

```diff
@@ -0,0 +1,119 @@
+import io.trino.plugin.iceberg.catalog.snowflake.IcebergSnowflakeCatalogConfig;
+import io.trino.plugin.iceberg.catalog.snowflake.TrinoSnowflakeCatalog;
+import io.trino.spi.catalog.CatalogName;
+import io.trino.spi.security.ConnectorIdentity;
+import io.trino.spi.type.TypeManager;
+import jakarta.annotation.PreDestroy;
+import org.apache.iceberg.IcebergBuild;
+import org.apache.iceberg.jdbc.JdbcClientPool;
+
+import java.net.URI;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+
+import static java.util.Objects.requireNonNull;
+import static org.apache.iceberg.jdbc.JdbcCatalog.PROPERTY_PREFIX;
+
+public class TrinoIcebergSnowflakeCatalogFactory
+        implements TrinoCatalogFactory
+{
+    // Below property values are copied from https://github.com/apache/iceberg/blob/main/snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java#L122-L129 (org.apache.iceberg:iceberg-snowflake:1.5.0)
+    private static final String JDBC_APPLICATION_PROPERTY = "application";
+    private static final String APP_IDENTIFIER = "iceberg-snowflake-catalog";
+    private static final String JDBC_USER_AGENT_SUFFIX_PROPERTY = "user_agent_suffix";
+    private static final int UNIQUE_ID_LENGTH = 20;
+
+    private final TrinoFileSystemFactory fileSystemFactory;
+    private final CatalogName catalogName;
+    private final TypeManager typeManager;
+    private final IcebergTableOperationsProvider tableOperationsProvider;
+
+    private final String snowflakeDatabase;
+    private final Map<String, String> snowflakeDriverProperties;
+    private final JdbcClientPool snowflakeConnectionPool;
+
+    @Inject
+    public TrinoIcebergSnowflakeCatalogFactory(
+            TrinoFileSystemFactory fileSystemFactory,
+            CatalogName catalogName,
+            IcebergSnowflakeCatalogConfig snowflakeCatalogConfig,
+            TypeManager typeManager,
+            IcebergTableOperationsProvider tableOperationsProvider)
+    {
+        this.fileSystemFactory = requireNonNull(fileSystemFactory, "fileSystemFactory is null");
+        this.catalogName = requireNonNull(catalogName, "catalogName is null");
+
+        this.snowflakeDriverProperties = getSnowflakeDriverProperties(
+                snowflakeCatalogConfig.getUri(),
+                snowflakeCatalogConfig.getUser(),
+                snowflakeCatalogConfig.getPassword(),
+                snowflakeCatalogConfig.getRole());
+        this.snowflakeDatabase = snowflakeCatalogConfig.getDatabase();
+        this.snowflakeConnectionPool = new JdbcClientPool(snowflakeCatalogConfig.getUri().toString(), snowflakeDriverProperties);
+
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        this.tableOperationsProvider = requireNonNull(tableOperationsProvider, "tableOperationsProvider is null");
+    }
+
+    @PreDestroy
+    public void shutdown()
+    {
+        snowflakeConnectionPool.close();
```

Comment:
Here the connection pool within the client is closed but there is a MetricsReporter which doesn't have any resource to close given its close method is unimplemented.

https://github.com/apache/iceberg/blob/apache-iceberg-1.5.0/snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java#L158-L159